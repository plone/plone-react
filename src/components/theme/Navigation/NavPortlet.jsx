import React from 'react';
import { List, Image } from 'semantic-ui-react';
import { Link as RouterLink } from 'react-router-dom';
import cx from 'classnames';
import { useDispatch, useSelector } from 'react-redux';
import { withRouter } from 'react-router';
import useDeepCompareEffect from 'use-deep-compare-effect';
import { getNavPortlet } from '@plone/volto/actions';
import { flattenToAppURL, getBaseUrl } from '@plone/volto/helpers';
import { Icon } from '@plone/volto/components';

import leftIcon from '@plone/volto/icons/left-key.svg';

function renderNode(node) {
  return (
    <List.Item key={node['@id']} active={node.is_current}>
      <List.Content>
        <RouterLink
          to={flattenToAppURL(node.href)}
          title={node.description}
          className={cx(`contenttype-${node.type}`, {
            in_path: node.is_in_path,
          })}
        >
          {node.thumb ? <Image src={flattenToAppURL(node.thumb)} /> : ''}
          {node.title}
          {node.is_current ? (
            <List.Content className="active-indicator">
              <Icon name={leftIcon} size="30px" />
            </List.Content>
          ) : (
            ''
          )}
        </RouterLink>

        {(node.items?.length && (
          <List.List>{node.items.map(renderNode)}</List.List>
        )) ||
          ''}
      </List.Content>
    </List.Item>
  );
}

/**
 * A navigation portlet implementation, similar to the classic Plone navigation
 * portlet. It uses the same API, so the options are similar to
 * INavigationPortlet
 *
 * @param {string} pathname Optional pathname to use for this portlet
 * @param {params} params Options for the navigation portlet
 *
 * name - The title of the navigation tree.
 * root_path - Root node path, can be "frontend path", derived from router
 * includeTop - Bool, Include top nodeschema
 * currentFolderOnly - Bool, Only show the contents of the current folder.
 * topLevel - Int, Start level
 * bottomLevel - Int, Navigation tree depth
 * no_icons - Bool, Suppress Icons
 * thumb_scale - String, Override thumb scale
 * no_thumbs = Bool, Suppress thumbs
 *
 */
function NavPortlet(props) {
  const {
    location,
    pathname = getBaseUrl(location.pathname),
    params = {},
  } = props;

  let qs = Object.keys(params)
    .sort()
    .map((key) => `expand.navportlet.${key}=${params[key]}`)
    .join('&');
  const path = `${pathname}${pathname.endsWith('/') ? '' : '/'}@navportlet${
    qs ? `?${qs}` : ''
  }`;

  const dispatch = useDispatch();
  const portlet = useSelector((state) => state.navPortlet?.[path]?.data);

  const { items = [] } = portlet || {};

  useDeepCompareEffect(() => {
    dispatch(getNavPortlet(pathname, params));
  }, [pathname, dispatch, params]);

  return items.length ? (
    <div className="navigation-portlet">
      {portlet.navigationportlet?.has_custom_name ? (
        <div className="nav-portlet-header">
          <RouterLink
            to={flattenToAppURL(portlet.navigationportlet?.url || '')}
          >
            {portlet.navigationportlet.title}
          </RouterLink>
        </div>
      ) : (
        ''
      )}
      <List>{items.map(renderNode)}</List>
    </div>
  ) : (
    ''
  );
}

export default withRouter(NavPortlet);
