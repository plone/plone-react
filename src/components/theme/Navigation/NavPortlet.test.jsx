import React from 'react';
import renderer from 'react-test-renderer';
import configureStore from 'redux-mock-store';
import { Provider } from 'react-intl-redux';
import { MemoryRouter } from 'react-router-dom';

import NavPortlet from './NavPortlet';

const mockStore = configureStore();

jest.mock('~/config', () => ({
  settings: {
    nonContentRoutes: [],
    supportedLanguages: ['en'],
    navDepth: 1,
  },
}));

describe('NavPortlet', () => {
  it('renders a navigation portlet component without active items', () => {
    const store = mockStore({
      navPortlet: {
        '/@navportlet': {
          data: {
            '@id': 'http://localhost:8080/Plone/@navportlet',
            title: 'Navigation',
            items: [
              {
                '@id': '/front-page',
                title: 'Welcome to Plone!',
                href: '/front-page',
              },
              { '@id': '/blog', title: 'Blog', href: '/blog' },
              { '@id': '/users', title: 'Users', href: '/users' },
            ],
          },
          error: undefined,
          loaded: true,
          loading: false,
        },
      },

      userSession: { token: '1234' },
      intl: {
        locale: 'en',
        messages: {},
      },
    });
    const component = renderer.create(
      <Provider store={store}>
        <MemoryRouter>
          <NavPortlet params={{}} url="/" />
        </MemoryRouter>
      </Provider>,
    );
    const json = component.toJSON();
    expect(json).toMatchSnapshot();
  });

  it('renders a navigation component with an active item', () => {
    const store = mockStore({
      navPortlet: {
        '/blog/@navportlet': {
          data: {
            '@id': 'http://localhost:8080/Plone/blog/@navportlet',
            title: 'Navigation',
            items: [
              {
                '@id': '/front-page',
                title: 'Welcome to Plone!',
                href: '/front-page',
              },
              {
                '@id': '/blog',
                title: 'Blog',
                href: '/blog',
                is_current: true,
                is_folderish: true,
                is_in_path: true,
              },
              { '@id': '/users', title: 'Users', href: '/users' },
            ],
          },
          error: undefined,
          loaded: true,
          loading: false,
        },
      },
      userSession: { token: '1234' },
      intl: {
        locale: 'en',
        messages: {},
      },
    });
    const component = renderer.create(
      <Provider store={store}>
        <MemoryRouter initialEntries={[{ pathname: '/blog' }]}>
          <NavPortlet pathname="/blog" />
        </MemoryRouter>
      </Provider>,
    );
    const json = component.toJSON();
    expect(json).toMatchSnapshot();
  });

  // it('renders a navigation component with an active item when its subchildren are accessed', () => {
  //   const store = mockStore({
  //     navigation: {
  //       items: [
  //         { title: 'Blog', url: '/blog' },
  //         { title: 'Users', url: '/users' },
  //       ],
  //     },
  //     userSession: { token: '1234' },
  //     intl: {
  //       locale: 'en',
  //       messages: {},
  //     },
  //   });
  //   const component = renderer.create(
  //     <Provider store={store}>
  //       <MemoryRouter initialEntries={[{ pathname: '/blog/2017/12/27' }]}>
  //         <Navigation pathname="/blog/2017/12/27" />
  //       </MemoryRouter>
  //     </Provider>,
  //   );
  //   const json = component.toJSON();
  //   expect(json).toMatchSnapshot();
  // });
  //
  // it('renders a navigation component with only one active item even if there are similar item names', () => {
  //   const store = mockStore({
  //     navigation: {
  //       items: [
  //         { title: 'Blog', url: '/blog' },
  //         { title: 'Blog of mine', url: '/blog-of-mine' },
  //         { title: 'Users', url: '/users' },
  //       ],
  //     },
  //     userSession: { token: '1234' },
  //     intl: {
  //       locale: 'en',
  //       messages: {},
  //     },
  //   });
  //   const component = renderer.create(
  //     <Provider store={store}>
  //       <MemoryRouter initialEntries={[{ pathname: '/blog' }]}>
  //         <Navigation pathname="/blog" />
  //       </MemoryRouter>
  //     </Provider>,
  //   );
  //   const json = component.toJSON();
  //   expect(json).toMatchSnapshot();
  // });
});
