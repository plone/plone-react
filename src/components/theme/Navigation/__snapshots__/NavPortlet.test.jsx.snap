// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`NavPortlet renders a navigation portlet component without active items 1`] = `
<div
  className="navigation-portlet"
>
  
  <div
    className="ui list"
    role="list"
  >
    <div
      className="item"
      onClick={[Function]}
      role="listitem"
    >
      <div
        className="content"
      >
        <a
          className="contenttype-undefined"
          href="/front-page"
          onClick={[Function]}
        >
          
          Welcome to Plone!
          
        </a>
        
      </div>
    </div>
    <div
      className="item"
      onClick={[Function]}
      role="listitem"
    >
      <div
        className="content"
      >
        <a
          className="contenttype-undefined"
          href="/blog"
          onClick={[Function]}
        >
          
          Blog
          
        </a>
        
      </div>
    </div>
    <div
      className="item"
      onClick={[Function]}
      role="listitem"
    >
      <div
        className="content"
      >
        <a
          className="contenttype-undefined"
          href="/users"
          onClick={[Function]}
        >
          
          Users
          
        </a>
        
      </div>
    </div>
  </div>
</div>
`;
